### PR TITLE
Added overload of WebApplication.CreateBuilder

### LINKED ISSUE #34837: Support specifying host configuration via WebApplication.CreateBuilder
## Background and Motivation

We've decided that it is not possible to mutate the host configuration (environment, application  name, content root) after the call to `WebApplication.Create`/`CreateBuilder` this simplifies the implementation as we don't need to handle the environment changing arbitrarily as [this issue](https://github.com/dotnet/aspnetcore/issues/34403) describes. Today the only way to change host configuration is to specify them in a command line args format which isn't very ergonomic. We should add an API that lets the user configure host configuration specifically as part of the Create/CreateBuilder call.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Builder
{
    public class WebApplication
    {
-       public static WebApplicationBuilder CreateBuilder(string[]? args = null);
+       public static WebApplicationBuilder CreateBuilder();
+       public static WebApplicationBuilder CreateBuilder(string[] args);
+       public static WebApplicationBuilder CreateBuilder(WebApplicationBuilderOptions options);
    }

+   public sealed class WebApplicationBuilderOptions
+   {
+       public string[]? Args { get; init; } 
+       public string? EnvironmentName { get; init; }
+       public string? ApplicationName { get; init; }
+       public string? ContentRootPath { get; init; }
+ }
}
```

## Usage Examples

```C#
var config = new WebApplicationOptions 
{
    Args = args,
    EnvironmentName = Environments.Staging,
    HostingStartupAssemblies = "Plugins",
};

var app = WebApplication.Create(options);

app.MapGet("/", (IHostEnvironment env) => env.EnvironmentName);

app.Run();
```

## Alternative Designs

```diff
namespace Microsoft.AspNetCore.Builder
{
    public class WebApplication
    {
-       public static WebApplication CreateBuilder(string[]? args = null);
+       public static WebApplicationBuilder CreateBuilder();
+       public static WebApplicationBuilder CreateBuilder(string[] args);
+       public static WebApplicationBuilder CreateBuilder(IConfiguration configuration);
    }
}
```

Args will win to allow hosts like WebApplicationFactory to override application behavior. This is important for testing.

``` C#
var config = new ConfigurationManager();
config.AddCommandLine(args);
config[HostDefaults.EnvironmentKey] = Environments.Development;
config[WebHostDefaults.HostingStartupAssembliesKey] = "Plugins";

var app = WebApplication.Create(config);

app.MapGet("/", (IHostEnvironment env) => env.EnvironmentName);

app.Run();
```

## Risks

None.

PS: I removed these options from `WebApplicationOptions` because they are super advanced and are typically controlled by the environment. We can always add them but it seems like it might be overkill here. Open to suggestions of course.

```
public string? HostingStartupAssemblies { get; init; }
public string? HostingStartupExcludedAssemblies { get; init; }
public bool? PreventHostingStartup { get; init; }
```

Also I only changed CreateBuilder, and not Create because this is a more advanced scenario.

### PR DESCRIPTION
**Pull request by davidfowl**
- Added WebApplicationOptions to that allow configuring any host configuration that can't be specified after the call.
- Added tests

Fixes #34837

PS: Seems like the public API analyzer is broken for me right now 

cc @captainsafia

### CODE REVIEW

File: src/DefaultBuilder/src/WebApplication.cs [hunk 1]
```diff
@@ -86,15 +86,23 @@ IServiceProvider IApplicationBuilder.ApplicationServices
         /// <param name="args">Command line arguments</param>
         /// <returns>The <see cref="WebApplication"/>.</returns>
         public static WebApplication Create(string[]? args = null) =>
-            new WebApplicationBuilder(Assembly.GetCallingAssembly(), args).Build();
+            new WebApplicationBuilder(Assembly.GetCallingAssembly(), new() { Args = args }).Build();
```
Comment (khellang) [inline src/DefaultBuilder/src/WebApplication.cs:89]: Y NO CHAINING? `Create(string[])` -> `CreateBuilder(string[])` + `Build()` -> `CreateBuilder(WebApplicationOptions)`

Comment (davidfowl) [reply src/DefaultBuilder/src/WebApplication.cs:89]: Because of `Assembly.GetCallingAssembly()` (which will be removed in another change)

Comment (khellang) [reply src/DefaultBuilder/src/WebApplication.cs:89]: Aaaah 💡

File: src/DefaultBuilder/src/WebApplicationBuilder.cs [hunk 2]
```diff
@@ -63,21 +65,26 @@ internal WebApplicationBuilder(Assembly? callingAssembly, string[]? args = null,
...
+
 
             Configuration = new();
 
             // This is the application configuration
             var hostContext = _bootstrapHostBuilder.RunDefaultCallbacks(Configuration, _hostBuilder);
 
             // Grab the WebHostBuilderContext from the property bag to use in the ConfigureWebHostBuilder
-            var webHostContext = (WebHostBuilderContext)_hostBuilder.Properties[typeof(WebHostBuilderContext)];
+            var webHostContext = (WebHostBuilderContext)hostContext.Properties[typeof(WebHostBuilderContext)];
```
Comment (captainsafia) [inline src/DefaultBuilder/src/WebApplicationBuilder.cs:87]: I think `_hostBuilder.Properties` and `hostContext.Properties` have the same data at this point. Why use one over the other?

Comment (davidfowl) [reply src/DefaultBuilder/src/WebApplicationBuilder.cs:87]: There was no point to this change, I just like using hostContext here as a result of this call `_bootstrapHostBuilder.RunDefaultCallbacks` call.

**Review by Tratcher — APPROVED** [review]

**Pull request merged**